Package Relay 1/3: Introduce TxDownloadManager and improve orphan-handling

glozow opened the pull request:
See #27463 for full project tracking.
Please see #27742 for how this PR fits into the big picture. This PR is based on a more recent commit than that one.

This branch includes:
(1) (now merged, see #30110) Introduces `TxDownloadManager`, which handles all transaction downloading. Adds tests for `TxDownloadManager`.
(2) (now merged, see #31397) Adds an "orphan resolution module". It adds all announcers of an orphan as potential resolution candidates, in a tracker implemented as a `TxRequestTracker`. In this PR, "orphan resolution" means requesting missing parents by `getdata(MSG_TX | MSG_WITNESS_FLAG, missing_txid)`. In a future PR, we'll add another resolution method, requesting ancestor wtxids using `getdata(MSG_ANCPKGINFO, orphan_wtxid)`.
(3) Makes `TxDownloadManager` internally thread-safe

glozow committed:
[txrequest] GetCandidatePeers and ResetRequestTimeout

glozow committed:
[txorphanage] support multiple announcers

Add ability to add and track multiple announcers per orphan transaction,
erasing announcers but not the entire orphan.

The tx creation code in orphanage_tests needs to be updated so that each
tx is unique, because the CountOrphans() check assumes that calling
EraseForPeer necessarily means its orphans are deleted.

glozow commented:
Investigating, thanks @DrahtBot

instagibbs commented on test/functional/p2p_segwit.py:
```diff
@@ -2058,7 +2058,7 @@ def received_wtxidrelay():
         test_transaction_acceptance(self.nodes[0], self.wtx_node, tx2, with_witness=True, accepted=False)
 
         # Expect a request for parent (tx) by txid despite use of WTX peer
-        self.wtx_node.wait_for_getdata([tx.sha256], 60)
```
can we programatically justify this magic :)

instagibbs reviewed (commented):
some initial comments through https://github.com/bitcoin/bitcoin/pull/28031/commits/116378efc1c9c1fe0d26cb42e2bdbb5770815c35

Log changes suggested are helpful for tracing what's happening in the orphanage on my node I'm testing.

dergoegge commented on src/txorphanage.cpp:
```diff
@@ -15,14 +14,35 @@ static constexpr int64_t ORPHAN_TX_EXPIRE_TIME = 20 * 60;
 bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer)
 {
     LOCK(m_mutex);
+    if (tx == nullptr) {
+        Assume(false);
+        return false;
+    }
```
Would be cleaner to write:
```suggestion
    if (!Assume(tx)) {
        return false;
    }
```

glozow replied:
Done

dergoegge commented on src/net_processing.cpp:
```diff
@@ -1434,6 +1441,49 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, const Peer& peer)
+void PeerManagerImpl::AddOrphanAnnouncer(NodeId nodeid, const uint256& orphan_wtxid, const CTransactionRef& tx, std::chrono::microseconds current_time)
+{
+    AssertLockHeld(::cs_main); // For m_txrequest
+    const bool connected = m_connman.ForNode(nodeid, [](CNode* node) { return node->fSuccessfullyConnected && !node->fDisconnect; });
+    if (!connected) return;
```
This can be dropped in favor of just checking that the CNodeState exists

glozow replied:
This is gone now, replaced with registration of `ConnectionInfo` when the node first connects.

dergoegge commented on src/net_processing.cpp:
```diff
@@ -1434,6 +1441,49 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, const Peer& peer)
+    AssertLockHeld(::cs_main); // For m_txrequest
+    const bool connected = m_connman.ForNode(nodeid, [](CNode* node) { return node->fSuccessfullyConnected && !node->fDisconnect; });
+    if (!connected) return;
+    if (m_txpackagetracker->Count(nodeid) + m_txrequest.Count(nodeid) >= MAX_PEER_TX_ANNOUNCEMENTS) {
```
(I've mentioned this offline as well)

The locking assumptions around this check are weird because `m_txpackagetracker` has its own internal mutex where as `m_txrequest` is guarded by `cs_main`. There is nothing stopping `m_txpackagetracker->Count()` form returning something different right after this check.

glozow replied:
Yeah good point.

It seems that, given the need to synchronize between `TxRequestTracker` and the package tracking stuff, we should have them both guarded by 1 lock. Looking at #26151, it seems like we could have a `m_txrequest GUARDED_BY(tx_download_mutex)`, `m_txpackagetracker GUARDED_BY(tx_download_mutex)`, and lock it from these peerman functions?

Alternatively, I wonder if it makes sense to take it a step further and put this all in a `TxDownloadManager` module that wraps orphanage, txrequest, and package tracking.

dergoegge commented on src/net_processing.cpp:
```diff
@@ -1434,6 +1441,49 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, const Peer& peer)
     }
 }
 
+void PeerManagerImpl::AddOrphanAnnouncer(NodeId nodeid, const uint256& orphan_wtxid, const CTransactionRef& tx, std::chrono::microseconds current_time)
```
```suggestion
void PeerManagerImpl::AddOrphanAnnouncer(NodeId nodeid, const CTransactionRef& tx, std::chrono::microseconds current_time)
```

dergoegge commented on src/txorphanage.cpp:
```diff
@@ -15,14 +14,35 @@ static constexpr int64_t ORPHAN_TX_EXPIRE_TIME = 20 * 60;
 /** Minimum time between orphan transactions expire time checks in seconds */
 static constexpr int64_t ORPHAN_TX_EXPIRE_INTERVAL = 5 * 60;
 
+void TxOrphanage::SubtractOrphanBytes(unsigned int size, NodeId peer)
```
Imo, it would be nicer to do the orphanage changes separately, including amended functional/unit/fuzz tests.

glozow replied:
> Imo, it would be nicer to do the orphanage changes separately, including amended functional/unit/fuzz tests.

Opened #28481 for orphanage changes

dergoegge commented on src/txorphanage.cpp:
```diff
@@ -15,14 +14,35 @@ static constexpr int64_t ORPHAN_TX_EXPIRE_TIME = 20 * 60;
+void TxOrphanage::SubtractOrphanBytes(unsigned int size, NodeId peer)
+{
+    AssertLockHeld(m_mutex);
+    Assume(m_peer_bytes_used.count(peer) > 0);
+    Assume(m_peer_bytes_used.at(peer) >= size);
```
Not a big fan of using `Assume` like this because we can't test these conditions but they **can** happen in production (if the caller is doing something wrong).

I'd suggest just returning if these assumptions don't hold.

dergoegge commented on src/node/txpackagetracker.h:
```diff
@@ -0,0 +1,88 @@
+namespace node {
+static constexpr bool DEFAULT_ENABLE_PACKAGE_RELAY{false};
+
+class TxPackageTracker {
```
Would be nice to add unit and fuzz tests for this

dergoegge commented on src/node/txpackagetracker.h:
```diff
@@ -0,0 +1,88 @@
+    /** Should be called when a transaction is accepted to the mempool. If it was an orphan we were
+     * trying to resolve, remove its entries from the orphanage and other data structures. If it is
+     * the ancestor of an orphan, add the orphan to its associated peer's workset. */
+    void MempoolAcceptedTx(const CTransactionRef& ptx);
```
```suggestion
    void MempoolAcceptedTx(const CTransactionRef& tx);
```

dergoegge commented on src/node/txpackagetracker.cpp:
```suggestion
            // Add the orphan's parents. Net processing will filter out what we already have.
            // Deduplicate parent txids, so that we don't have to loop over
            // the same parent txid more than once down below.
            std::set<uint256> unique_parents;
            auto to_prevout = [](const CTxIn& in) { return in.prevout.hash; };
            std::transform(ptx->vin.begin(), ptx->vin.end(), std::inserter(unique_parents, unique_parents.begin()), to_prevout);
```

dergoegge commented on src/node/txpackagetracker.cpp:
Would you mind adding some spacing around logical blocks in your code? I find a lot of the new code hard to read with a lack of spacing

e.g.:
```suggestion
    {
        AssertLockNotHeld(m_mutex);
        LOCK(m_mutex);

        std::vector<std::pair<NodeId, GenTxid>> expired;
        auto tracker_requestable = orphan_request_tracker.GetRequestable(nodeid, current_time, &expired);
        for (const auto& entry : expired) {
            LogPrint(BCLog::TXPACKAGES, "\nTimeout of inflight %s %s from peer=%d\n", entry.second.IsWtxid() ? "ancpkginfo" : "orphan parent",
                entry.second.GetHash().ToString(), entry.first);
        }

        std::vector<GenTxid> results;
        for (const auto& gtxid : tracker_requestable) {
```

mzumsande commented on src/node/txpackagetracker.cpp:
I don't think there should ever be a `\n` prefix, why would we want to separate the meta information (timestamp, threadname etc.) from the actual log entry?

mzumsande reviewed (commented):
some more comments, not finished yet though.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -2966,7 +2966,7 @@ bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
             RelayTransaction(orphanHash, porphanTx->GetWitnessHash());
             m_txpackagetracker->AddChildrenToWorkSet(*porphanTx);
-            m_txpackagetracker->EraseOrphanTx(orphanHash);
+            m_txpackagetracker->EraseOrphanTx(porphanTx->GetWitnessHash());
```
nit: could use `orphan_wtxid` introduced in the previous commit, here and in various other places.

mzumsande reviewed (commented):
Some thoughts about logging:

After testing this on mainnet for a bit, I think it would be nice to be able to follow the fate of each individual orphan in `BCLog::TXPACKAGES`:
We currently have a tx-level based log entry for addition (“stored orphan..."), but nothing for removal, so having additional log entries in `MempoolAcceptedTx()` for succesful resolution and `MempoolRejectedTx()` for failure would be nice, maybe also a tx-level message for removal due to expiration and overflow.

If this became too spammy, we could use different severities, but I think it shouldn't be too bad except if you just started with an empty mempool?

Also, could we log wtxids whenever possible? There is currently a bit of a mix.

mzumsande commented on src/node/txpackagetracker.cpp:
```diff
@@ -4,41 +4,157 @@
+        if (tx == nullptr && !m_orphanage.HaveTx(GenTxid::Wtxid(wtxid))) return;
+
+        // Even though this stores the orphan wtxid, GenTxid::Txid instead of Wtxid because we will be requesting the parents via txid.
+        orphan_request_tracker.ReceivedInv(nodeid, GenTxid::Txid(wtxid), is_preferred, reqtime);
```
(moved over from [27742](https://github.com/bitcoin/bitcoin/pull/27742#discussion_r1265739284)):
The approach to add a wtxid disguised as a txid to distinguish between legacy orphan processing and package relay seems a bit like a hack to me.
I'm not strictly against it, but I guess I don't completely understand yet why it's necessary - the information whether a peer supports package relay does not change, so why can't we just always use `GenTxid::Wtxid(wtxid)` and check again in `GetOrphanRequests()` and elsewhere whether we do package relay with the peer instead of checking whether it's a txid / wtxid?

glozow replied:
Ah good point, it's not necessary, we could just use `GenTxid::Wtxid` for all of them. :+1:

instagibbs commented on src/net_processing.cpp:
```diff
@@ -5926,6 +5958,22 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
+                // We don't know that the transaction was rejected by mempool. But if the
+                // transaction was added to mempool, we would have already called
+                // MempoolAcceptedTx().
+                m_txpackagetracker->MempoolRejectedTx(gtxid.GetHash());
```
also, adding wtxid logging at `MEMPOOLREJ` log is done would be <3

glozow replied:
Added `MEMPOOLREJ` log and successful `AcceptToMemoryPool` log for both orphan and non-orphan tx results.

instagibbs commented on src/node/txpackagetracker.cpp:
This logging should be handled by the caller, since the caller sometimes isn't actually requesting the items!

glozow committed:
[txorphanage] store parent txids in OrphanTx

glozow commented:
Opened #28199 with tests that may help us ensure this PR isn't changing specific behaviors. Marking this as draft as it depends on that one.

glozow commented on src/node/txpackagetracker.cpp:
```diff
@@ -0,0 +1,181 @@
+    {
+        LOCK(m_mutex);
+        m_orphanage.AddChildrenToWorkSet(*ptx);
+        m_orphanage.EraseTx(ptx->GetWitnessHash());
```
Thanks, fixed

glozow commented on src/node/txpackagetracker.cpp:
```diff
@@ -0,0 +1,181 @@
+    void MempoolRejectedTx(const uint256& wtxid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+        m_orphanage.EraseTx(wtxid);
```
Fixed

glozow commented on src/txorphanage.cpp:
```diff
@@ -15,6 +14,16 @@ static constexpr int64_t ORPHAN_TX_EXPIRE_TIME = 20 * 60;
+{
+    AssertLockHeld(m_mutex);
+    Assume(m_peer_bytes_used.count(peer) > 0);
+    Assume(m_peer_bytes_used.at(peer) >= size);
```
Changed these to `if(!Assume(...)) return;`

glozow commented on src/node/txpackagetracker.h:
```diff
@@ -59,6 +58,30 @@ class TxPackageTracker {
+
+    /** Number of packages we are working on with this peer. Includes any entries in the orphan
+     * tracker and in-flight requests. */
+    size_t Count(NodeId nodeid) const;
```
Replaced the exposure of `Count` to `CheckIsEmpty()` functions

glozow commented on src/net_processing.cpp:
```diff
@@ -2960,9 +2960,10 @@ bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
-            LogPrint(BCLog::MEMPOOL, "   accepted orphan tx %s\n", orphanHash.ToString());
+            LogPrint(BCLog::MEMPOOL, "   accepted orphan tx %s\n", orphan_wtxid.ToString());
```
Logging both :+1:

glozow commented on src/txorphanage.h:
```diff
@@ -61,7 +62,31 @@ class TxOrphanage {
+
+    /** Remove a peer from an orphan's announcers list, erasing the orphan if this is the only peer
+     * who announced it. If the orphan doesn't exist or does not list this peer as an announcer, do nothing. */
+    void EraseOrphanOfPeer(const uint256& wtxid, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
Deleted from that commit

glozow commented on src/net_processing.cpp:
```diff
@@ -4286,6 +4285,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                 if (RecursiveDynamicUsage(*ptx) < 100000) {
                     AddToCompactExtraTransactions(ptx);
                 }
+                m_txpackagetracker->MempoolRejectedTx(tx.GetWitnessHash());
```
Rearranged the commits so it's more clear where the erasure is new

glozow commented on src/node/txpackagetracker.cpp:
```diff
@@ -4,41 +4,157 @@
 
 namespace node {
+    /** How long to wait before requesting orphan ancpkginfo/parents from an additional peer. */
+    static constexpr auto ORPHAN_ANCESTOR_GETDATA_INTERVAL{60s};
```
fixed

glozow reviewed (commented):
Reworked this into a `TxDownloadManager` (https://github.com/bitcoin/bitcoin/pull/28031#discussion_r1262316909) and addressed most comments.

glozow commented on src/node/txpackagetracker.cpp:
```diff
@@ -4,41 +4,157 @@
+        }
+        for (const auto& wtxid : wtxids_erased) {
+            if (block_wtxids.count(wtxid) == 0) {
+                conflicted_wtxids.insert(wtxid);
```
Removed the special casing. I can't remember why it comes into play later, but if I do, I'll add it when it's needed.

glozow commented:
> Some thoughts about logging:
> 
> After testing this on mainnet for a bit, I think it would be nice to be able to follow the fate of each individual orphan in `BCLog::TXPACKAGES`: We currently have a tx-level based log entry for addition (“stored orphan..."), but nothing for removal, so having additional log entries in `MempoolAcceptedTx()` for succesful resolution and `MempoolRejectedTx()` for failure would be nice, maybe also a tx-level message for removal due to expiration and overflow.
> 
> If this became too spammy, we could use different severities, but I think it shouldn't be too bad except if you just started with an empty mempool?
> 
> Also, could we log wtxids whenever possible? There is currently a bit of a mix.

> also, adding wtxid logging at `MEMPOOLREJ` log is done would be <3

@mzumsande @instagibbs
I've split the commits adding new logs and wtxids to the logging to its own small PR, #28364

(Also, rebased.)

instagibbs commented on src/node/txdownload_impl.cpp:
```diff
@@ -0,0 +1,415 @@
+            m_orphan_resolution_tracker.RequestedTx(nodeid, orphan_gtxid.GetHash(),
+                                                    current_time + ORPHAN_ANCESTOR_GETDATA_INTERVAL);
+        } else {
+            LogPrint(BCLog::TXPACKAGES, "couldn't find parent txids to resolve orphan %s with peer=%d\n",
```
under what circumstances is this expected? Eviction of the orphan in the orphanage, but outstanding resolution request? A comment to the expected conditions would be helpful for documentation.

instagibbs replied:
ok, it's happening because of: https://github.com/bitcoin/bitcoin/pull/28031/files#r1309300862

instagibbs reviewed (commented):
went over new-er implementation through 7f9db92ba3f483f78bd248406d921bd73b1e7054

dropping some comments

fuzzer fail:
```
node/txdownload_impl.cpp:18 ConnectedPeer: Assertion `m_peer_info.count(nodeid) == 0' failed.
```

instagibbs commented on src/node/txdownload_impl.cpp:
```diff
@@ -0,0 +1,415 @@
+    EXCLUSIVE_LOCKS_REQUIRED(m_tx_download_mutex)
+{
+    if (!Assume(m_peer_info.count(peer) > 0)) return;
+    if (m_orphanage.HaveTx(gtxid)) AddOrphanAnnouncer(peer, gtxid.GetHash(), now);
```
this adds an announcer even if the INV is txid type, even though the assumption is wtxid arg

instagibbs replied:
I think the simplest solution is just filter this addition based on wtxid-ness, since this only helps in the multi-announcer case, which wouldn't be a regression, iiuc

glozow replied:
ooh thanks :+1:

glozow replied:
Fixed, just did a gate on IsWtxid. Don't think that many non-wtxidrelay nodes are still out there anyway.

instagibbs commented on src/node/txdownload_impl.cpp:
```diff
@@ -0,0 +1,415 @@
+            if (!Assume(m_peer_info.count(nodeid) > 0)) continue;
+            const auto& info = m_peer_info.at(nodeid).m_connection_info;
+            for (const auto& txid : *parent_txids) {
+                // Schedule with no delay. It should be requested immediately
```
```suggestion
                // Schedule with no delay, i.e. not using ReceivedTxInv(). It should be requested immediately
```

glozow replied:
Added

instagibbs commented:
think this can be rebased since tests are merged

glozow commented:
Rebased but not all comments addressed yet, still working on it.

glozow commented on src/txorphanage.h:
```diff
@@ -101,6 +104,8 @@ class TxOrphanage {
         int64_t nTimeExpire;
         size_t list_pos;
         std::set<NodeId> announcers;
+        /** Txids of the missing parents to request. Determined by peerman. */
```
Yes

glozow commented on src/txorphanage.h:
```diff
@@ -45,13 +45,13 @@ class TxOrphanage {
     /** Maybe erase all orphans announced by a peer (eg, after that peer disconnects). If an orphan
      * has been announced by another peer, don't erase, just remove this peer from the list of announcers. */
-    void EraseForPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+    std::vector<uint256> EraseForPeer(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
Removed, I think in a past implementation I was using it but I'm not anymore

glozow commented on src/txorphanage.cpp:
```diff
@@ -266,3 +299,24 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
+{
+    AssertLockNotHeld(m_mutex);
+    LOCK(m_mutex);
+    // Nothing to do if this peer isn't storing any orphans.
```
You're right, removed in #28481

glozow commented on src/txorphanage.cpp:
```diff
@@ -226,7 +234,7 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
             if (itByPrev == m_outpoint_to_orphan_it.end()) continue;
             for (auto mi = itByPrev->second.begin(); mi != itByPrev->second.end(); ++mi) {
                 const CTransaction& orphanTx = *(*mi)->second.tx;
-                const uint256& orphanHash = orphanTx.GetHash();
```
Renamed to `orphan_wtxid` in #28481

glozow commented on src/txorphanage.h:
```diff
@@ -61,7 +61,28 @@ class TxOrphanage {
+        LOCK(m_mutex);
+        return m_total_orphan_bytes;
+    }
+    /** Return total amount of orphans stored by this transaction, in bytes. */
```
haha, fixed in #28481

glozow commented on src/net_processing.cpp:
```diff
@@ -2954,54 +3052,15 @@ bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
+                state.ToString());
+            // Ignoring the return value. Within orphan processing, we do not make
+            // additional orphan resolution requests when a transaction is missing inputs.
+            ProcessInvalidTx(porphanTx, peer.m_id, state);
```
Yes, behavior change. To make more concrete:
1. We get an orphan C with 2 missing parents, A and B
2. We reject A
3. We get B and accept it
4. We add C to workset and process it in `ProcessOrphanTx`.
5. Here, since we find A in recent_rejects, we also add C to recent_rejects when we wouldn't have before.

I think this is good actually... :thinking:

instagibbs replied:
IIRC I was fine with it, just making sure I understood!

glozow commented on src/node/txdownload_impl.cpp:
```diff
@@ -22,22 +25,30 @@ void TxDownloadImpl::DisconnectedPeer(NodeId nodeid)
 {
     LOCK(m_tx_download_mutex);
-    m_orphanage.EraseForBlock(block);
+    const auto erased_wtxids = m_orphanage.EraseForBlock(block);
```
Done

glozow commented on src/node/txdownload_impl.cpp:
```diff
@@ -216,6 +230,39 @@ std::vector<GenTxid> TxDownloadImpl::GetRequestsToSend(NodeId nodeid, std::chron
+    const auto orphans_ready = m_orphan_resolution_tracker.GetRequestable(nodeid, current_time, &expired_orphan_resolution);
+    // Expire orphan resolution attempts
+    for (const auto& [nodeid, orphan_gtxid] : expired_orphan_resolution) {
+        LogPrintf("timeout of in-flight orphan resolution %s for peer=%d\n", orphan_gtxid.GetHash().ToString(), nodeid);
```
oop changed

glozow commented on src/node/txdownload_impl.cpp:
```diff
@@ -216,6 +230,39 @@ std::vector<GenTxid> TxDownloadImpl::GetRequestsToSend(NodeId nodeid, std::chron
+                                                    current_time + ORPHAN_ANCESTOR_GETDATA_INTERVAL);
+        } else {
+            LogPrint(BCLog::TXPACKAGES, "couldn't find parent txids to resolve orphan %s with peer=%d\n",
+                     nodeid, orphan_gtxid.GetHash().ToString());
```
Fixed

glozow commented on src/node/txdownloadman.h:
```diff
@@ -68,6 +68,13 @@ class TxDownloadManager {
     /** Should be called when a notfound for a tx has been received. */
     void ReceivedNotFound(NodeId nodeid, const std::vector<uint256>& txhashes) { m_impl->ReceivedNotFound(nodeid, txhashes); }
+
+    /** Add a new orphan transaction. Returns whether this orphan is going to be processed and the
```
added

glozow commented on src/txrequest.h:
```diff
@@ -195,6 +201,9 @@ class TxRequestTracker {
     size_t Size() const;
 
+    /** For some tx hash (either txid or wtxid), return all peers with non-COMPLETED announcements. */
+    std::vector<NodeId> GetCandidatePeers(const uint256& txhash) const;
```
Probably not?

glozow commented on src/net_processing.cpp:
```diff
@@ -690,8 +688,7 @@ class PeerManagerImpl final : public PeerManager
     CTxMemPool& m_mempool;
 
     /** Protects tx download, rejection filter. */
```
Removed comment

glozow commented on src/net_processing.cpp:
```diff
@@ -4016,39 +4016,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-                }
-
-                if (m_txdownloadman.GetOrphanageRef().AddTx(ptx, pfrom.GetId(), unique_parents)) {
+                const auto& [will_process_orphan, unique_parents] = m_txdownloadman.NewOrphanTx(ptx, pfrom.GetId(), current_time);
```
Yep. We only want to `AddToCompactExtraTransactions` if this is a new orphan that we're actually keeping. We might fail to add it to orphanage if it's full.

instagibbs commented on src/txorphanage.cpp:
```diff
@@ -138,32 +212,38 @@ void TxOrphanage::LimitOrphans(unsigned int max_orphans)
+        const auto& wtxid = m_orphan_list[randompos]->second.tx->GetWitnessHash();
+        expired_and_evicted.emplace_back(wtxid);
+        EraseTxNoLock(wtxid);
         ++nEvicted;
```
we should `+=` like the above call or do `++` for both I think

glozow replied:
done

instagibbs commented on src/txorphanage.cpp:
```diff
@@ -226,18 +317,64 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
-        for (const uint256& orphanHash : vOrphanErase) {
-            nErased += EraseTxNoLock(orphanHash);
+        for (const uint256& orphan_wtxid : vOrphanErase) {
+            nErased += EraseTxNoLock(orphan_wtxid);
```
```
echo Kv//////////////lJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSU//////////////////////////////+A////////kEEKJf//ev///////8P///////8AAIH//wkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkB/wD8///SQP///////////////////////////////////4D///////+QQQol//96////////w/////////8KJf//ev//////////kEEKJf//1v///wP/////kEH///+QQQo1//8BPwAAv67Kf/////8u//////////////////////////////////////////////////////////////////////////96AP/1///////S0P//IP96AP////////+A////////kEEKJf//ev//////////kEEKJf//1v///wP/////kEH///+QQQo1//96/////P////+QQQol/yX//3r//////////wol///W//////////97e3t7e3t7e3t7e3t7e3t7e3t7e/+QQQol//96//////////////8s/yb/ | base64 -d > crash.txt
```

These call `EraseTxNoLock` without doing m_peer_work_set management:
TxOrphanage::EraseForBlock
TxOrphanage::LimitOrphans

This results in fuzzer crashes where it's (still) in the peer's work set, but not counted as an announcer since it's been evicted(then later sent).

Any time `EraseTx{NoLock}` is called we need to make sure that `m_peer_work_set` is appropriately being cleaned.

Two other spots I haven't thought through yet but I suspect need the same fix:
TxDownloadImpl::MempoolAcceptedTx
TxDownloadImpl::MempoolRejectedTx

glozow replied:
Is it a problem to lazily handle transactions in peer work sets that have been deleted from the orphanage? It's cleaner to `remove_work_from_all_sets` of course, but it iterates through every entry in `m_peer_work_set` so there is a performance difference but the external outcome is the same.

I think erasing proactively in `EraseOrphanOfPeer` is necessary because the tx isn't deleted. Trying to avoid returning the tx in `GetTxToReconsider(peer)` after we've already decided to give up on (tx, peer).

glozow replied:
> TxOrphanage::EraseForBlock
> TxOrphanage::LimitOrphans

> This results in fuzzer crashes where it's (still) in the peer's work set, but not counted as an announcer since it's been evicted(then later sent).

Question: is it crashing because the tx is being returned from `GetTxToReconsider`? Seems weird as it should return a nullptr if it gets a tx that isn't there anymore...

OOhhhhhhhhh could it be that it's re-added? My guess:
1. `AddTx(tx_child, peer)`
2. `AddChildrenToWorkSet(tx_parent)` -> peer workset contains tx_child
3. `EraseTx(tx_child)` -> peer workset still contains tx_child but we expect orphanage to lazily delete that when we call `GetTxToReconsider`
4. `AddTx(tx_child, other_peer)` (makes no sense but we're fuzzing)
5. `GetTxToReconsider(peer)` returns `tx_child` even though it's not in `orphan_resolution_tracker`

instagibbs replied:
> Is it a problem to lazily handle transactions in peer work sets that have been deleted from the orphanage?

Since `GetTxToReconsider` appears to be used in a single spot, and it really doesn't matter that it once was announced, then erased, then reconsidered for a peer, I suppose the fuzz test should just be modified to not crash :)

glozow replied:
Yeah. Removed the `Assert(HaveTxAndPeer())` bit

glozow replied:
Removed `Assert(HaveTxAndPeer())` in the fuzzer, and added comments for when we clean up the work sets and when we don't

instagibbs reviewed (commented):
Some not snapshot-ready fixes for fuzz crashes I worked through:

https://github.com/instagibbs/bitcoin/commit/5388830c3c5bb0ed9a1d1379ec94b285de68aa9a

glozow commented on src/test/fuzz/txorphan.cpp:
```diff
@@ -91,8 +93,11 @@ FUZZ_TARGET(txorphan, .init = initialize_orphanage)
-                            bool have_tx = orphanage.HaveTx(GenTxid::Txid(ref->GetHash())) || orphanage.HaveTx(GenTxid::Wtxid(ref->GetHash()));
-                            Assert(have_tx);
+                            Assert(orphanage.HaveTx(GenTxid::Txid(ref->GetHash())) ||
+                                   orphanage.HaveTx(GenTxid::Wtxid(ref->GetHash())));
```
hahahahaha
Unfortunately #28107 doesn't help in this case since this function takes a gentxid, but I've rebased on top of that PR.

glozow commented on src/test/fuzz/txdownloadman.cpp:
```diff
@@ -0,0 +1,406 @@
+                if (ptx) {
+                    // However, if there was a non-null tx in the workset, HaveMoreWork should have
+                    // returned true.
+                    Assert(expect_work);
```
an entry in `orphan_resolution_tracker`?

instagibbs replied:
Was going to suggest adding a check for `OrphanTx.announcers`, but it may not actually exist as discussed elsewhere.

glozow replied:
That, and we can't check `announcers` because it's private

glozow commented:
Note I know there's a "needs rebase" label but that's because I rebased on top of 28107

glozow committed:
[refactor] change type of unique_parents to Txid

glozow committed:
[txorphanage] return wtxids from EraseForBlock

glozow committed:
[p2p] try multiple peers for orphan resolution

glozow committed:
[txorphanage] return erased wtxids from LimitOrphans

glozow committed:
[functional test] orphan handling with multiple announcers

glozow committed:
[unit test] multiple announcers

glozow committed:
[fuzz] orphanage multiple announcer functions

glozow committed:
[p2p] only attempt 1p1c when both txns provided by the same peer

Now that we track all announcers of an orphan, it's not helpful to
consider an orphan provided by a peer that didn't send us this parent.
It can only hurt our chances of finding the right orphan when there are
multiple candidates.

Adapt the 2 tests in p2p_opportunistic_1p1c.py that looked at 1p1c
packages from different peers. Instead of checking that the right peer
is punished, we now check that the package is not submitted. We can't
use the functional test to see that the package was not considered
because the behavior is indistinguishable (except for the logs).

glozow committed:
[p2p] grab m_tx_download_mutex only when needed

The places where m_tx_download_mutex is held in between calls to
txdownloadman are unnecessary.

For example, m_tx_download_mutex was previously held throughout these steps:
  1. ReceivedTx(), receive should_validate=true
  2. MempoolRejectedTx(tx), receive package_to_validate={parent, child}
  3. MempoolRejectedPackage(parent, child)
  4. MempoolRejectedTx(parent)
  5. MempoolRejectedTx(child)

But this is not necessary for correctness. If, for example,
BlockConnected is called in between any of the steps, txdownloadman's
internal state will remain consistent, even if the relevant
transaction(s) confirmed. See the txdownloadman fuzzers.

This sets up the next commit which moves locking into TxDownloadManagerImpl.

glozow committed:
add internal Mutex to TxDownloadManagerImpl

It is named slightly differently to make it easy to check that
m_tx_download_mutex is fully removed in the next commit.

glozow committed:
remove m_tx_download_mutex in net_processing

glozow committed:
[refactor] split AlreadyHaveTx into internal and external versions

glozow committed:
[txdownload] remove unique_parents that we already have

This means we no longer return parents we already have in the
m_unique_parents result from MempoolRejectedTx.

We need to separate the loop that checks AlreadyHave parents from the
loop that adds parents as announcments, because we may do the latter
loop multiple times for different peers.

glozow committed:
[refactor] make Find1P1CPackage internal

glozow commented:
Rebased. I have split the first 12 commits (the majority of this PR) into #31397. The other 5 commits are for making the module internally thread-safe, and can be opened in another PR afterward.

glozow commented:
Closing because this is pretty completed (yay!). TxDownloadManager was added in #30110, and the multi-announcer orphan handling was added in #31397. We just need the last 2 commits to make txdownloadman internally thread-safe, but I believe we have somebody working on it.
